libraries/AP_ExternalAHRS/AP_ExternalAHRS_GSOF.cpp

```diff
@@ -0,0 +1,362 @@
+                    case AP_GSOF::GnssStatus::GNSS_SPS_MODE: // TODO Is this right?
+                        gps_data.fix_type = AP_GPS_FixType::NONE;
+                        break;
+                    case AP_GSOF::GnssStatus::DGPS_SPS_MODE:
+                        gps_data.fix_type = AP_GPS_FixType::DGPS;
+                        break;
+                    case AP_GSOF::GnssStatus::GNSS_PPS_MODE:
+                        gps_data.fix_type = AP_GPS_FixType::PPP;
+                        break;
+                    case AP_GSOF::GnssStatus::FIXED_RTK_MODE:
+                        gps_data.fix_type = AP_GPS_FixType::RTK_FIXED;
+                        break;
+                    case AP_GSOF::GnssStatus::FLOAT_RTK_MODE:
+                        gps_data.fix_type = AP_GPS_FixType::RTK_FLOAT;
+                        break;
+                    case AP_GSOF::GnssStatus::DR_MODE:
+                        gps_data.fix_type = AP_GPS_FixType::NONE;
+                        break;
+                }
+                if (ins_full_nav.gnss_status != AP_GSOF::GnssStatus::FIX_NOT_AVAILABLE) {
+                    // If fix is unavailble, the lat/lon may be zero, so don't post the data.
+                    // To reproduce this condition, disconnect the GPS antenna and reboot the receiver.
+                    post_filter();
+                }
+            }
+            if (parsed.get(AP_GSOF::INS_RMS)) {
+                last_ins_rms_ms = now;
+
+                gps_data.horizontal_pos_accuracy = Vector2d(ins_rms.pos_rms_n, ins_rms.pos_rms_e).length();
+                gps_data.vertical_pos_accuracy = ins_rms.pos_rms_d;
+                gps_data.horizontal_vel_accuracy = Vector2d(ins_rms.vel_rms_n, ins_rms.vel_rms_e).length();
+            }
+            if (parsed.get(AP_GSOF::LLH_MSL)) {
+                last_llh_msl_ms = now;
+
+                gps_data.longitude = static_cast<int32_t>(llh_msl.longitude * 1E7);
+                gps_data.latitude = static_cast<int32_t>(llh_msl.latitude * 1E7);
+                gps_data.msl_altitude = static_cast<int32_t>(llh_msl.altitude_msl * 1E2);
+            }
+
+            if (parsed.get(AP_GSOF::LLH_MSL) && parsed.get(AP_GSOF::INS_FULL_NAV)) {
+                undulation = ins_full_nav.altitude - llh_msl.altitude_msl;
+            }
+
+            uint8_t instance;
+            AP_GSOF::MsgTypes expected_gps;
+            expected_gps.set(AP_GSOF::POS_TIME);
+            expected_gps.set(AP_GSOF::INS_FULL_NAV);
+            expected_gps.set(AP_GSOF::INS_RMS);
+            expected_gps.set(AP_GSOF::LLH_MSL);
+            if (AP::gps().get_first_external_instance(instance) && parsed == expected_gps) {
+                AP::gps().handle_external(gps_data, instance);
+            }
+        }
+        hal.scheduler->delay_microseconds(100);
+        check_initialise_state();
+    }
+}
+
+void AP_ExternalAHRS_GSOF::check_initialise_state(void)
+{
+    const auto new_init_state = initialised();
```

Comment:
```suggestion
    const bool new_init_state = initialised();
```

You can take the use of `auto` too far :-)